### ISSUE TITLE
window.onload() inside renderValue

### ISSUE THREAD

**Issue by SymbolixAU** [ramnathv/htmlwidgets#223]
**TL;DR**

Is there any reason why putting a `window.onload = function() {  <create objects & run code> }` inside `renderValue: function(x) {  window.onload  }` would only work intermittently in Shiny/Rstudio Viewer?

---

In trying to implement my [google maps widget](https://github.com/SymbolixAU/googleway/blob/master/inst/htmlwidgets/google_map.js) I've been having intermittent success. 

To use the Google Maps Javascript API you need to call the source JS via a url, and call a callback function

```
<script src="https://maps.googleapis.com/maps/api/js?key=YOUR_API_KEY&callback=initMap"
    async defer></script>
```

As per the suggestions you gave me in issue #212 I had this working by passing the script call as a dependency ([complete google_map.R](https://github.com/SymbolixAU/googleway/blob/master/R/google_map.R) )

```
googlemap$dependencies <- c(googlemap$dependencies,
                              list(
                                htmltools::htmlDependency(
                                  name = "googleway",
                                  version = "9999",
                                  src=".",
                                  head = paste0('<script src="https://maps.googleapis.com/maps/api/js?key=', key, '&libraries=visualization"></script>'),
                                  all_files = FALSE
                                  )
                                )
                              )
```

This dependency needs to load before I create my `map` object, so I put it inside a `window.onload()` function ([full google_map.js code](https://github.com/SymbolixAU/googleway/blob/master/inst/htmlwidgets/google_map.js) )

```
renderValue: function(x) {
    window.onload = function() {
          var mapDiv = document.getElementById(el.id);
          mapDiv.className = "googlemap";
          var map = new google.maps.Map(mapDiv, {
            center: {lat: x.lat, lng: x.lng},
            zoom: x.zoom
          });
```

When I first implemented this everything worked - in RStudio Viewer and Shiny - but, the more I test and run it I'm finding `onload()` only works sometimes in Shiny (by checking the JS Console log in Chrome). (It's also not showing in the Rstudio Viewer, but it does render when I select 'show in new window' from Rstudio). 

Do you have any suggestions as to what might be causing this; or advice on a different approach to delay the creation of the `map` object until after the JS dependency has loaded?

**Comment by SymbolixAU** [comment]
One approach might be to use `setInterval()` to continually check for the resource. This *seems* to work (for shiny). However, I'm not happy about this as a solution (it seems very in-elegant).

```
 var checkExists = setInterval(function(){

            var map = new google.maps.Map(mapDiv, {
              center: {lat: x.lat, lng: x.lng},
              zoom: x.zoom
            });

            if (google !== 'undefined'){
              console.log("exists");
              clearInterval(checkExists);
            }else{
              console.log("does not exist!");
            }
           // do stuff - load layers etc

```

**Comment by jcheng5** [comment]
Huh, I would've expected it never to work, as `window.onload` should have fired long before `renderValue` runs...?

**Comment by SymbolixAU** [comment]
maybe I occasionally got lucky and it loaded the google JS resource before it tried to create the map object?

**Comment by SymbolixAU** [comment]
The solution above works for when loading the widget in a browser, but not for the Rstudio Viewer pane. Is there a check I can do within `renderValue` to check what mode it's in, something like:

```
 if (HTMLWidgets.shinyMode){
    // load widget in browser
}else if (in Rstudio Viewer){
   // load widget in Rstudio
}
```

**Comment by jjallaire** [comment]
When running in the Viewer the query param viewer_pane=1 is appended to the
URL.
